src/DotNetEnv/Parsers.cs

```diff
@@ -173,17 +173,23 @@ private static string ToUtf32Char (string hex)
                 .Or(OctalChar)
                 .Or(EscapedChar);
 
+        internal static readonly Parser<IEnumerable<char>> InlineCommentBeginOrControl =
+            Parse.Regex("[ \t]#").Or(Parse.Char(c => char.IsControl(c) && c != '\t', "controls except tab").Once());
+
         // unquoted values can have interpolated variables,
         // but only inline whitespace -- until a comment,
         // and no escaped chars, nor byte code chars
-        // FIXME: would be nice to solve multi word with comment case with parser directly (no split + trim)
         internal static readonly Parser<ValueCalculator> UnquotedValue =
-            InterpolatedValue.Or(
-                NotControlNorWhitespace("'\"$")
-                .Or(InlineWhitespaceChars.AtLeastOnce().Text())
-                .AtLeastOnce()
-                .Select(strs => new ValueActual(strs))
-            ).Many().Select(vs => new ValueCalculator(vs).Split("[ \t]#").Trim());
+            InterpolatedValue
+                .Or(Parse.CharExcept('$').Except(InlineCommentBeginOrControl)
+                    .AtLeastOnce()
+                    .Text()
+                    .Select(x => new ValueActual(x))
+                )
+                .Many()
+                .Except(Parse.Regex("[ \t]*['\"#]"))
```

Comment:
Except after Many resolves to an initial check whether we directly begin with an unmatched quote or a comment.

The following Or assures, that we never fail in here and return empty values if nothing matches. Invalid contents are handled by Assignment then.